modules/queue/queue_redis.go

```diff
@@ -0,0 +1,222 @@
+		return nil, errors.New("no redis host found")
+	} else if len(dbs) == 1 {
+		queue.client = redis.NewClient(&redis.Options{
+			Network:  config.Network,
+			Addr:     strings.TrimSpace(dbs[0]), // use default Addr
+			Password: config.Password,           // no password set
+			DB:       config.DBIndex,            // use default DB
+		})
+	} else {
+		queue.client = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs: dbs,
+		})
+	}
+	if err := queue.client.Ping().Err(); err != nil {
+		return nil, err
+	}
+	queue.pool.qid = GetManager().Add(queue, RedisQueueType, config, exemplar, queue.pool)
+
+	return queue, nil
+}
+
+// Run runs the redis queue
+func (r *RedisQueue) Run(atShutdown, atTerminate func(context.Context, func())) {
+	atShutdown(context.Background(), r.Shutdown)
+	atTerminate(context.Background(), r.Terminate)
+
+	go func() {
+		_ = r.pool.AddWorkers(r.workers, 0)
+	}()
+
+	go r.readToChan()
+
+	log.Trace("RedisQueue: %s Waiting til closed", r.name)
+	<-r.closed
+	log.Trace("RedisQueue: %s Waiting til done", r.name)
+	r.pool.Wait()
+
+	log.Trace("RedisQueue: %s Waiting til cleaned", r.name)
+	ctx, cancel := context.WithCancel(context.Background())
+	atTerminate(ctx, cancel)
+	r.pool.CleanUp(ctx)
+	cancel()
+}
+
+func (r *RedisQueue) readToChan() {
+	for {
+		select {
+		case <-r.closed:
+			// tell the pool to shutdown
+			r.pool.cancel()
+			return
+		default:
+			bs, err := r.client.LPop(r.queueName).Bytes()
+			if err != nil && err != redis.Nil {
+				log.Error("RedisQueue: %s Error on LPop: %v", r.name, err)
+				time.Sleep(time.Millisecond * 100)
+				continue
+			}
+
+			if len(bs) == 0 {
+				time.Sleep(time.Millisecond * 100)
+				continue
+			}
+
+			var data Data
+			if r.exemplar != nil {
+				t := reflect.TypeOf(r.exemplar)
+				n := reflect.New(t)
+				ne := n.Elem()
+				err = json.Unmarshal(bs, ne.Addr().Interface())
+				data = ne.Interface().(Data)
+			} else {
+				err = json.Unmarshal(bs, &data)
+			}
+			if err != nil {
+				log.Error("RedisQueue: %s Error on Unmarshal: %v", r.name, err)
+				time.Sleep(time.Millisecond * 100)
+				continue
+			}
+
+			log.Trace("RedisQueue: %s Task found: %#v", r.name, data)
+			r.pool.Push(data)
+			time.Sleep(time.Millisecond * 10)
+		}
+	}
+}
+
+// Push implements Queue
+func (r *RedisQueue) Push(data Data) error {
+	if r.exemplar != nil {
+		// Assert data is of same type as r.exemplar
+		value := reflect.ValueOf(data)
+		t := value.Type()
+		exemplarType := reflect.ValueOf(r.exemplar).Type()
+		if !t.AssignableTo(exemplarType) || data == nil {
+			return fmt.Errorf("Unable to assign data: %v to same type as exemplar: %v in %s", data, r.exemplar, r.name)
+		}
+	}
+	bs, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	return r.client.RPush(r.queueName, bs).Err()
+}
+
+// Shutdown processing from this queue
+func (r *RedisQueue) Shutdown() {
+	log.Trace("Shutdown: %s", r.name)
+	select {
+	case <-r.closed:
+	default:
+		close(r.closed)
```

Comment:
Is it possible that two goroutines will attempt to call `Shutdown()` at the same time? If that's the case, there's a slight chance that `close()` can be called twice if both threads pass `<-r.closed` at the same time.

Reply:
I'll add a mutex to both this and the ledis disk queue.